src/main.cc

```diff
@@ -1536,6 +1536,9 @@ SquidMain(int argc, char **argv)
 
 #endif
 
+    Mem::Init();
```

Comment:
I pushed it before as a matter of optimization: AnyP::UriScheme uses SBuf, which benefits from MemPools being available.
Mem::Init might be called on demand from elsewhere as well, but then why is it in SquidMain at all?

Reply:
> I pushed it before as a matter of optimization: AnyP::UriScheme uses SBuf, which benefits from MemPools being available.

Understood, thank you. MemPools are _naturally_ used (including used for SBufs!) even _before_ main() is called:

```
#0  Mem::Init () at old_api.cc:282
#1  0x0000555555eae185 in GetPool (type=1) at old_api.cc:70
#2  0x0000555555eae74c in memAllocate (type=MEM_32B_BUF) at old_api.cc:124
#3  0x0000555555eae9a9 in memAllocBuf (net_size=0, gross_size=0x7fffffffe2e0) at old_api.cc:197
#4  0x0000555555e11829 in MemBlob::memAlloc (this=0x5555562526e0, minSize=0) at MemBlob.cc:94
#5  0x0000555555e10e71 in MemBlob::MemBlob (this=0x5555562526e0, reserveSize=0, __in_chrg=<optimized out>,  __vtt_parm=<optimized out>) at MemBlob.cc:57
#6  0x0000555555e12d56 in SBuf::GetStorePrototype () at SBuf.cc:78
#7  0x0000555555e12ade in SBuf::SBuf (this=0x7fffffffe460, S=0x555555ed80a8 "public") at SBuf.cc:60
#8  0x00005555559fd4fd in LookupTable<HttpHdrCcType, LookupTableRecord<HttpHdrCcType>, CaseInsensitiveSBufHash>::LookupTable (this=0x555556166ea0 <ccLookupTable>, theInvalid=CC_OTHER, data=0x5555560ec0e0 <CcAttrs>) at base/LookupTable.h:61
#9  0x00005555559fcad0 in __static_initialization_and_destruction_0 (__initialize_p=1, __priority=65535) at HttpHdrCc.cc:47
#10 0x00005555559fcb85 in _GLOBAL__sub_I_ccLookupTable () at HttpHdrCc.cc:352
#11 0x00007ffff71faebb in call_init (env=<optimized out>, argv=0x7fffffffe558, argc=7) at ../csu/libc-start.c:145
#12 __libc_start_main_impl (main=0x555555b342cc <main(int, char**)>, argc=7, argv=0x7fffffffe558, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe548) at ../csu/libc-start.c:379
#13 0x00005555559684a5 in _start ()
```

AFAICT, attempts to move Mem::Init() call somewhere (within/after main()) in hope to make MemPools "being available" earlier are both futile/harmful and outside this PR scope. In other words, MemPools should be treated as always available, and any changes that may be interpreted as contradicting that principle are unwanted.

Please do not touch/move Mem::Init() call in this PR.

> Mem::Init might be called on demand from elsewhere as well, but then why is it in SquidMain at all?

Problematic Squid code exists for several reasons, but none of those reasons are related to (a polished version of) this PR. FWIW, I would welcome a dedicated high-quality PR removing all these wasteful "explicit" Mem::Init() calls and, hence, preventing similar future doubts/discussions. The `pools` static inside GetPool() is already following the "no globals" principle, and GetPool() already calls `pools` initialization function -- Mem::Init(). All that is left for use to do is to hide that initialization function from public API (e.g., by inlining it into GetPool()). None of that is related to bug 5337 and this PR though.

Reply:
The C++ MEMPROXY pooling APIs will initialize as-needed. `Mem::Init()` is part of the legacy APIs and is not strictly necessary in `SquidMain()` today - removal has not been tested, but likely fine.

Reply:
> `Mem::Init()` ... is not strictly necessary in `SquidMain()` today - removal has not been tested, but likely fine

... _and_ is outside this PR scope, especially since there are other Mem::Init() calls that should be removed (and the possibility of new calls should be prevented!) at the same time.